sway-core/src/parse_tree/code_block.rs

```diff
@@ -58,55 +58,111 @@ impl CodeBlock {
                         warnings,
                         errors
                     );
-                    vec![AstNode {
-                        content: AstNodeContent::Expression(evaluated_node),
+                    let mut ast_node_contents = evaluated_node_result
+                        .var_decls
+                        .into_iter()
+                        .map(|x| AstNode {
+                            content: AstNodeContent::Declaration(Declaration::VariableDeclaration(
+                                x,
+                            )),
+                            span: span::Span {
+                                span: pair.as_span(),
+                                path: path.clone(),
+                            },
+                        })
+                        .collect::<Vec<_>>();
+                    ast_node_contents.push(AstNode {
+                        content: AstNodeContent::Expression(evaluated_node_result.value),
                         span: span::Span {
                             span: pair.as_span(),
                             path: path.clone(),
                         },
-                    }]
+                    });
+                    ast_node_contents
                 }
                 Rule::return_statement => {
-                    let evaluated_node = check!(
+                    let res_result = check!(
                         ReturnStatement::parse_from_pair(pair.clone(), config),
                         continue,
                         warnings,
                         errors
                     );
-                    vec![AstNode {
-                        content: AstNodeContent::ReturnStatement(evaluated_node),
+                    let mut ast_node_contents = res_result
+                        .var_decls
+                        .into_iter()
+                        .map(|x| AstNode {
+                            content: AstNodeContent::Declaration(Declaration::VariableDeclaration(
```

Comment:
does this mean return statements must necessarily return a variable declaration? I'm not sure I understand this bit

Reply:
Because match expressions now have the ability to create AstNodes, those nodes need to be inserted before the return statement node (the same goes for other nodes that are parsed). This bit of code is taking those new AstNodes and inserting them.